go/fix-chain/fixer.go

```diff
@@ -0,0 +1,282 @@
+			//log.Printf("fetch issuer %d from %s", i, url)
+			ferr := augmentIntermediates(fix.opts.Intermediates,
+				url, fix.fixer.cache)
+			if ferr != nil {
+				ferr.Cert = fix.cert
+				ferr.Chain = fix.chain.certs
+				fix.fixer.errors <- ferr
+			}
+			chain, err := fix.cert.Verify(*fix.opts)
+			if err == nil {
+				//dumpChains("fixed", chain)
+				fix.fixer.fixed++
+				fix.fixer.log.PostChains(chain)
+				return
+			}
+		}
+	}
+	//log.Printf("failed to fix certificate for %s", fix.cert.Subject.CommonName)
+	fix.fixer.notfixed++
+	fix.fixer.errors <- &FixError{Type: FixFailed, Cert: fix.cert,
+		Chain: fix.chain.certs}
+}
+
+type Fixer struct {
+	fix    chan *Fix
+	active uint32
+	// Counters may not be entirely accurate due to non-atomicity
+	skipped          uint
+	reconstructed    uint
+	notreconstructed uint
+	fixed            uint
+	notfixed         uint
+	alreadydone      uint
+
+	wg     sync.WaitGroup
+	log    *Log
+	errors chan *FixError
+	cache  *URLCache
+	done   map[[HashSize]byte]bool
+}
+
+func (f *Fixer) fixChain(cert *x509.Certificate, d *DedupedChain, intermediates *x509.CertPool, l *Log) {
+	if l.Posted(cert) {
+		//log.Printf("Skip already posted cert")
+		f.skipped++
+		return
+	}
+	opts := x509.VerifyOptions{Intermediates: intermediates,
+		Roots: l.Roots(), DisableTimeChecks: true,
+		KeyUsages: []x509.ExtKeyUsage{x509.ExtKeyUsageAny}}
+	chain, err := cert.Verify(opts)
+	if err == nil {
+		//dumpChains("verified", chain)
+		f.reconstructed++
+		l.PostChains(chain)
+		return
+	}
+	//log.Printf("failed to verify certificate for %s: %s", cert.Subject.CommonName, err)
+	f.errors <- &FixError{Type: VerifyFailed, Cert: cert, Chain: d.certs,
+		Error: err}
+	f.notreconstructed++
+	f.deferFixChain(cert, d, &opts)
+}
+
+func (f *Fixer) FixAll(d *DedupedChain) {
+	h := HashBag(d.certs)
+	if f.done[h] {
+		f.alreadydone++
+		return
+	}
+	f.done[h] = true
+	intermediates := x509.NewCertPool()
+	for _, c := range d.certs {
+		intermediates.AddCert(c)
+	}
+	for _, c := range d.certs {
+		f.fixChain(c, d, intermediates, f.log)
+	}
+}
+
+func (f *Fixer) deferFixChain(cert *x509.Certificate, chain *DedupedChain,
+	opts *x509.VerifyOptions) {
+	f.fix <- &Fix{cert: cert, chain: chain, opts: opts, fixer: f}
+}
+
+func (f *Fixer) fixServer() {
+	defer f.wg.Done()
+
+	for fix := range f.fix {
+		atomic.AddUint32(&f.active, 1)
+		fixChain(fix)
+		atomic.AddUint32(&f.active, ^uint32(0))
+	}
+}
+
+func (f *Fixer) Wait() {
+	close(f.fix)
+
+	// Must wait for fixers first, in case they log something.
+	f.wg.Wait()
+	f.log.Wait()
+}
+
+func NewFixer(logurl string, errors chan *FixError) *Fixer {
+	f := &Fixer{fix: make(chan *Fix), log: NewLog(logurl, errors),
+		errors: errors, cache: NewURLCache(),
+		done: make(map[[HashSize]byte]bool)}
+
+	for i := 0; i < 100; i++ {
+		f.wg.Add(1)
+		go f.fixServer()
+	}
+
+	t := time.NewTicker(time.Second)
+	go func() {
+		for _ = range t.C {
+			log.Printf("fixers: %d active, %d skipped, " +
```

Comment:
This disagrees with `gofmt`. I'm not sure, but I think the Go dogma is that there is no line length limit, so you just keep going.

One part that I *am* sure is that you **never** win an argument with `gofmt`. :wink: